src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs

```diff
@@ -10,6 +10,7 @@ namespace System.Text.Json.Nodes
 {
     public partial class JsonObject : IDictionary<string, JsonNode?>
     {
+        private readonly int? _initialCapacity;
```

Comment:
It seems that this additional field could be avoided?

Reply:
Can you say a bit more about that? I am working under the assumption that we don't want to change the lazy-initialized `Dictionary`, which means (I _think_) that we have to be able to hold on to that known initial capacity until we actually do the initialization. Are you saying we should change the initialization logic? For instance, something like go ahead and initialize the dictionary in the constructor if we know the capacity?

Reply:
Let me know if https://github.com/dotnet/runtime/pull/96486/commits/782f43b41b7973f8e8df4d6f214aff6080de80cd is along the right direction 🙏

Reply:
I think this could be simplified a bit. If you take look at what `InitializeDictionary()` does is that it checks if the `_dictionary` field has been created and otherwise populates with values from the `JsonElement` it might contain.

I think the constructor could be improved a lot if you just populated the dictionary directly, roughly:
```C#
JsonPropertyDictionary<..> dictionary = properties is ICollection<...> collection ? new(collection.Count) : new();

foreach (KeyValuePair<...> entry in properties)
{
    dictionary.Add(entry.Key, entry.Value);
}

_dictionary = dictionary; // write the populated dictionary to the field
```

Reply:
I think https://github.com/dotnet/runtime/pull/96486/commits/09b6f40ec030483fa8f60ec3af8c7f80c791a7b8 should be moving in the right direction.

This change has an (unfortunate?) side effect of having to copy the `value?.AssignParent` line from the existing `Add` method (because it depends on `Dictionary`). Let me know if you are ok with that!